### PR TITLE
Issues/25

### LINKED ISSUE #25: Add static code analysis to the build
Add jacoco, checkstyle, pmd, etc.

### PR DESCRIPTION
**Pull request by nickgrealy**
Fixes #25 

Changes proposed in this pull request:
- Added static code analysis tools - jacoco, checkstyle, pmd, jdepend
- Bumped LTS version to 2.7.4
- [N/A] Includes tests covering the new functionality?
- [x] Ready for review
- [x] Follows CONTRIBUTING rules

@nickgrealy

### CODE REVIEW

**Comment by nickgrealy** [comment]
@Jimilian - some pretty big changes, can you please review when you get a spare minute?

File: config/gradle/common.gradle [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+apply plugin: 'java'
+
+idea.project?.vcs = 'Git'
+
+// --- Setup script specific properties ---
+
+ext {
+    /* optional - set variable enforceJavaVersion */
+    enforceJavaVersionFlag = project.getProperties().containsKey('enforceJavaVersion')
+    javaVersion = project.getProperties().get('enforceJavaVersion') ?: JavaVersion.VERSION_1_6
```
Comment (Jimilian) [inline config/gradle/common.gradle]: Why JavaVersion.VERSION_1_6? I thought Jenkins supports 1_7 only.

File: config/gradle/common.gradle [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        delete ".gradle"
+    }
+
+    tasks.matching { it instanceof JavaCompile || it instanceof GroovyCompile }.all {
+        options.compilerArgs += ["-Xlint:unchecked", "-Xlint:deprecation"]
+        options.fork = true
+        options.forkOptions.jvmArgs += customJvmArgs
+    }
+
+//    tasks.withType(JavaCompile) {
```
Comment (Jimilian) [inline config/gradle/common.gradle]: Some debug code? I think it's better to remove commented stuff before publishing.

File: src/main/groovy/org/jenkinsci/plugins/globalEventsPlugin/GlobalComputerListener.java [hunk 3]
```diff
@@ -5,17 +5,17 @@
...
 import java.util.logging.Logger;
+
 /**
  * Warning: This MUST stay a Java class, Groovy cannot compile (for some reason??).
  */
 @Extension
 public class GlobalComputerListener extends ComputerListener {
 
-    protected static Logger log = Logger.getLogger(GlobalComputerListener.class.getName());
+    private static Logger log = Logger.getLogger(GlobalComputerListener.class.getName());
```
Comment (Jimilian) [inline src/main/groovy/org/jenkinsci/plugins/globalEventsPlugin/GlobalComputerListener.java]: `final`?

File: src/main/groovy/org/jenkinsci/plugins/globalEventsPlugin/GlobalEventsPlugin.groovy [hunk 4]
```diff
@@ -230,12 +238,12 @@ class GlobalEventsPlugin extends Plugin implements Describable<GlobalEventsPlugi
          * @param groovyCode
          * @param params
          */
-        private FormValidation safeExecGroovyCode(
+        protected FormValidation safeExecGroovyCode(
```
Comment (Jimilian) [inline src/main/groovy/org/jenkinsci/plugins/globalEventsPlugin/GlobalEventsPlugin.groovy]: Why do we need `protected`?

File: src/test/groovy/org/jenkinsci/plugins/globalEventsPlugin/acceptance/StepDefs.groovy [hunk 5]
```diff
@@ -5,12 +5,7 @@ import cucumber.api.java.en.Given
 import cucumber.api.java.en.Then
 import cucumber.api.java.en.When
 import hudson.util.FormValidation
-import org.jenkinsci.plugins.globalEventsPlugin.GlobalEventsPlugin
-import org.jenkinsci.plugins.globalEventsPlugin.GlobalEventsPluginTest
-import org.jenkinsci.plugins.globalEventsPlugin.GlobalRunListener
-import org.jenkinsci.plugins.globalEventsPlugin.GlobalComputerListener
-import org.jenkinsci.plugins.globalEventsPlugin.GlobalQueueListener
-import org.jenkinsci.plugins.globalEventsPlugin.LoggerTrap
+import org.jenkinsci.plugins.globalEventsPlugin.*
```
Comment (Jimilian) [inline src/test/groovy/org/jenkinsci/plugins/globalEventsPlugin/acceptance/StepDefs.groovy]: I didn't get `import *` policy. In some places you changed `*` to set of imports, here you changes several imports to one asterisk :)

**Comment by Jimilian** [comment]
Hi, @nickgrealy! I hope that everything is ok :) Are you going to finish this change?

**Comment by markyjackson-taulia** [comment]
No longer valid

**Pull request closed without merging**